### PR TITLE
[Feature] Adds a rocm profiler script to profile kernels using rocprofv3

### PR DESCRIPTION
**Pull request by diptorupd**
### Summary

Adds a reusable ROCm profiling library and a first driver benchmark (FA2 single-prefill) with a
single, fully-automated pipeline to generate bth timings and roofline analysis.

- **New module** `flashinfer/rocm_profiler/rocm_profiler.py` — generic `RocmProfiler` /
  `KernelConfig` library; works with any callable kernel.
- **New benchmark** `flashinfer/benchmarks/rocm_benchmarks/bench_fa2_prefill.py` — thin
  driver sweeping 9 FA2 prefill configs across the MI300X roofline ridge point.
- All generated outputs (`*.csv`, `*_counters.yml`, `*_roofline.png`) are gitignored.

### What the profiler does

1. **Timing pass** — repeated GPU launches with configurable dry-run and measurement windows;
   writes `{label}_timing.csv`.
2. **Counter collection** — spawns one `rocprofv3` subprocess *per config* (solves the
   ~8-dispatch ring-buffer limit). Each subprocess re-enters the driver script via the
   `_ROCM_PROFILER_INTERNAL` env sentinel and runs only a single profiled dispatch.
   Multi-pass CSVs are merged into one wide-form row per config; writes
   `{label}_counter_collection.csv`.
3. **Roofline plot** — two-panel log-log plot (TFLOPS + TB/s vs arithmetic intensity) with
   MI300X hardware ceilings, MFMA utilisation annotation, and LDS bank-conflict overlay;
   writes `{label}_roofline.png`.

### Key design decisions

| Decision | Rationale |
|---|---|
| Per-config `rocprofv3` subprocess | `rocprofv3` has a ≈8-dispatch ring buffer; a full sweep overwrites early configs |
| JIT pre-warm subprocess before profiling | FlashInfer JIT compilation deadlocks inside `rocprofv3`'s instrumented environment |
| Two-pass counter splits (`FetchSize` / `WriteSize`) | gfx942 hw constraint — sharing the same internal resource causes error code 38 |
| `MFMA_util%` derived metric | `VALUBusy` is unavailable; MFMA utilisation is computed from collected MFMA op counts + kernel duration |

### Using a custom counter preset

```yaml
# my_counters.yml  (rocprofv3 native job format)
jobs:
  # Pass 1 — counters that can share hardware resources
  - pmc: [SQ_WAVES, SQ_INSTS_MFMA, FetchSize]
  # Pass 2 — counters that conflict with Pass 1 (split required on gfx942)
  - pmc: [SQ_WAIT_INST_ANY, WriteSize]

### CODE REVIEW

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+from pathlib import Path
+
+import torch
+import flashinfer
+from flashinfer.jit.core import logger
+
+logger.setLevel(logging.ERROR)
+
+# rocm_profiler lives at flashinfer/rocm_profiler/rocm_profiler.py
+# (two levels up from benchmarks/rocm_benchmarks/)
```
Comment (demandal25) [inline benchmarks/rocm_benchmarks/bench_fa2_prefill.py]: I don't think we need such comments!

**Review by demandal25 — COMMENTED** [review]
I am assuming agent made the changes and it's impossible to review such changes. 
- How do I know if the implementation is correct?
- Can you please address the copilot comments?

Further, how do we use it to compare with the upstream? Do we still use the upstream benchmarking scripts for that and use this script for iterative development of rocm?

**Review by demandal25 — APPROVED** [review]
Thanks for the PR!

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            )
+        )
+    return configs
+
+
+# Constructed at module level so rocprofv3 subprocess replay (which re-imports
+# this module with the same sys.argv) picks up the correct counters preset.
+profiler = RocmProfiler(
+    configs=_make_configs(),
+    num_warmup=3,
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py]: Fixed

File: rocm_profiler/rocm_profiler.py [hunk 3]
```diff
@@ -0,0 +1,1412 @@
...
+            f"[rocm_profiler] JIT pre-warm (LD_LIBRARY_PATH={prewarm_env.get('LD_LIBRARY_PATH', 'unchanged')}) ..."
+        )
+        sys.stdout.flush()
+        pre = subprocess.run(
+            [sys.executable, sys.argv[0]],
+            env=prewarm_env,
+            capture_output=True,
+            text=True,
+            timeout=300,
+        )
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply rocm_profiler/rocm_profiler.py:677]: Fixed

File: rocm_profiler/rocm_profiler.py [hunk 4]
```diff
@@ -0,0 +1,1412 @@
...
+                    "--output-format",
+                    "csv",
+                    "--output-directory",
+                    tmpdir,
+                    "--output-file",
+                    self.label,
+                ]
+                if self.kernel_name_regex:
+                    cmd += ["--kernel-include-regex", self.kernel_name_regex]
+                cmd += ["--", sys.executable, sys.argv[0]]
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply rocm_profiler/rocm_profiler.py]: Done

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py (hunk 1 above)
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py]: Removed

[the remark this answers is not in the document]
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py]: Removed the comment

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        causal_str = "causal" if causal else "nc"
+        configs.append(
+            KernelConfig(
+                name=f"s{seq_len}_{causal_str}",
+                run_fn=lambda q=q, k=k, v=v, c=causal: (
+                    flashinfer.single_prefill_with_kv_cache_return_lse(
+                        q, k, v, causal=c, backend="fa2"
+                    )
+                ),
+                theoretical_flops=flops,
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py:182]: Fixed

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        v = torch.randn(
+            seq_len, num_kv_heads, head_dim, dtype=torch.half, device="cuda"
+        )
+
+        # FLOPs: Q·Kᵀ + softmax(S)·V, causal mask halves each matmul
+        factor = 2 if causal else 4
+        flops = seq_len * seq_len * num_qo_heads * head_dim * factor
+
+        # Bytes: Q + K + V + O  (4 tensors × fp16 × elements), cold-cache lower bound
+        theo_bytes = 8 * seq_len * num_qo_heads * head_dim * 2
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py]: Fixed

File: rocm_profiler/rocm_profiler.py [hunk 7]
```diff
@@ -0,0 +1,1412 @@
...
+                }
+
+                try:
+                    result = subprocess.run(
+                        cmd,
+                        capture_output=True,
+                        text=True,
+                        env=env,
+                        timeout=self._rocprofv3_timeout,
+                    )
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply rocm_profiler/rocm_profiler.py:729]: Fixed

File: rocm_profiler/rocm_profiler.py [hunk 8]
```diff
@@ -0,0 +1,1412 @@
...
+    """
+    rocprofv3 --output-format csv produces long-form data:
+      Each dispatch appears once per counter, with Counter_Name / Counter_Value columns.
+
+    This function:
+      1. Groups all rows by Dispatch_Id.
+      2. Pivots Counter_Name / Counter_Value into columns.
+      3. Filters to only the profiled dispatches listed in dispatch_map.
+      4. Annotates each row with config_name and writes wide-form CSV to out_csv.
+
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply rocm_profiler/rocm_profiler.py]: cleaned up

File: benchmarks/rocm_benchmarks/bench_fa2_prefill.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        num_warmup=3,
+        dry_run_ms=100,
+        repeat_ms=1000,
+        counters=_counters,
+        kernel_name_regex="SinglePrefillWithKVCacheKernel",
+        output_dir=_OUTPUT_DIR,
+        label=_label,
+        roofline=(_counters == "roofline"),
+    )
+    profiler.run()
```
[the remark this answers is not in the document]
Comment (diptorupd) [reply benchmarks/rocm_benchmarks/bench_fa2_prefill.py:204]: fixed
